### PR TITLE
Issue85 bestmatchattribute

### PR DESCRIPTION
**Pull request by doutriaux1**
@danlipsa @aashish24 @dnadeau4 this is ready for review

it does best match suggestion to user when the user sends the wrong attribute

```python
In [1]: import vcs

In [2]: b = vcs.createboxfill()

In [3]: b.datawc_X1 = .5
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-b30909989229> in <module>()
----> 1 b.datawc_X1 = .5

/git/vcs/vcs/__init__.pyc in __setattr__(self, a, v)
     68                 raise AttributeError(
     69                     "'%s' object has no attribute '%s' did you mean one of %s ?" %
---> 70                     (self.__class__.__name__, a, repr(real_matches)))
     71             else:
     72                 raise AttributeError(

AttributeError: 'Gfb' object has no attribute 'datawc_X1' did you mean one of ['datawc_y1', 'datawc_x1'] ?

In [4]: b.leggend = 'gh'
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-adae8dab4c80> in <module>()
----> 1 b.leggend = 'gh'

/git/vcs/vcs/__init__.pyc in __setattr__(self, a, v)
     64                 raise AttributeError(
     65                     "'%s' object has no attribute '%s' did you mean %s ?" %
---> 66                     (self.__class__.__name__, a, repr(real_matches[0])))
     67             elif len(real_matches) > 1:
     68                 raise AttributeError(

AttributeError: 'Gfb' object has no attribute 'leggend' did you mean 'legend' ?

In [5]: b.randomstuffhere = 'randomstuffhere'
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-6-4cf0a312baf0> in <module>()
----> 1 b.randomstuffhere = 'randomstuffhere'

/git/vcs/vcs/__init__.pyc in __setattr__(self, a, v)
     73                     "'%s' object has no attribute '%s' valid attributes are: %s" %
     74                     (self.__class__.__name__, a, repr(
---> 75                         self.__slots__)))
     76 
     77 

AttributeError: 'Gfb' object has no attribute 'randomstuffhere' valid attributes are: ['colormap', '_colormap', '__doc__', 'name', 'g_name', 'xaxisconvert', 'yaxisconvert', 'levels', 'fillareacolors', 'fillareastyle', 'fillareaindices', 'fillareaopacity', 'fillareapixelspacing', 'fillareapixelscale', 'ext_1', 'ext_2', 'missing', 'projection', 'xticlabels1', 'xticlabels2', 'yticlabels1', 'yticlabels2', 'xmtics1', 'xmtics2', 'ymtics1', 'ymtics2', 'datawc_x1', 'datawc_x2', 'datawc_y1', 'datawc_y2', 'boxfill_type', 'color_1', 'color_2', 'level_1', 'level_2', 'legend', 'boxfill_type', 'datawc_timeunits', 'datawc_calendar', '_name', '_xaxisconvert', '_yaxisconvert', '_levels', '_fillareacolors', '_fillareastyle', '_fillareaindices', '_fillareaopacity', '_fillareapixelspacing', '_fillareapixelscale', '_ext_1', '_ext_2', '_missing', '_projection', '_xticlabels1', '_xticlabels2', '_yticlabels1', '_yticlabels2', '_xmtics1', '_xmtics2', '_ymtics1', '_ymtics2', '_datawc_x1', '_datawc_x2', '_datawc_y1', '_datawc_y2', '_legend', '_boxfill_type', '_color_1', '_color_2', '_level_1', '_level_2', '_datawc_timeunits', '_datawc_calendar', 'info']

In [6]: 
```

### CODE REVIEW

**Comment by doutriaux1** [comment]
@aashish24 @danlipsa does your silence mean you approve it? @dnadeau4 please review as well because I would like to see this going into cdms.

**Comment by danlipsa** [comment]
@doutriaux1 So if you misspell an attribute, you'll get a suggestion with the correct spelling? That's pretty cool! How well does it work? Does it slow dow processing when all attributes are correct?

**Comment by doutriaux1** [comment]
@danlipsa I don't think it slows down much because in that case the only added test 
```python
if value in self.__slots__:
   blah
```

**Comment by doutriaux1** [comment]
blah is the same as before.

File: vcs/Canvas.py [hunk 1]
```diff
@@ -349,12 +349,13 @@ class Canvas(object):
...
         'backend',
         'configurator',
+        '_Canvas__last_plot_actual_args',
+        '_Canvas__last_plot_keyargs',
         '__last_plot_actual_args',
         '__last_plot_keyargs',
         '_continents',
         '_continents_line',
         '_savedcontinentstype',
-        '__weakref__',
```
Comment (doutriaux1) [inline vcs/Canvas.py]: @danlipsa we might want to put this one back in? Do we?

**Comment by danlipsa** [comment]
> if value in self.__slots__:

@doutriaux1 This is a bunch of string comparisons for each attribute. Is there any way you can do this only on the exception generated when an attribute was not found.

**Comment by doutriaux1** [comment]
@danlipsa it's a good idea I will `try` to stick it inside a `try/except` call.

**Comment by doutriaux1** [comment]
@danlipsa I just tried this and it's not possible, if we have the __setattr__ then it seems to overwrite the `object` class `__slots__` check and we can set any attribute we want. So I can't put it in a `try/except` statement. In any case I suspect the `object` class does the test anyway. So I think we're safe and we are not adding a double check for the attribute being in `__slots__`

**Comment by doutriaux1** [comment]
@danlipsa while I was at it I also added it to a simple getattr running the test suite now and will push if it didn't break anything.

File: vcs/Canvas.py (hunk 1 above)
Comment (danlipsa) [inline vcs/Canvas.py]: grep-ing through sources seems that we only hold a weak reference to a VTKVCSBackend object. So it seems safe to delete this one.

**Comment by danlipsa** [comment]
@doutriaux1 My concern with this change is that it is going to slow down plotting. I know we have name based array access in VTK and when we optimize a loop, this is often the place to look.
I think it would be worth to run some tests with and without this change to see the impact.

In my mind, this very cool when you get your program working, after that it should be disabled. If the impact is big, maybe a way to still add this is make all checks only if a boolean such as EnableBestMatchAttribute is set to true.

**Comment by doutriaux1** [comment]
@danlipsa I disagree on this the impact is minimal. Again before it was inheriting from `object` class which DOES the check to see if it's in `__slots__` but now because we redefine `__setattr__` that check gets bypassed. So it's a wash. And it is really really nice for the end user. Think of `git` for example which tells you `did you mean` if you misspell anything.

**Comment by doutriaux1** [comment]
@danlipsa I can do some timing if you want to see if it adds anything.

**Comment by danlipsa** [comment]
@doutriaux1 Thanks. I like it as well, if it does not add much in running time costs.
Seems like a lot to do a bunch of string comparisons for each attribute set. Maybe python does some sort of hash. I am curious about the test results.

**Comment by doutriaux1** [comment]
@danlipsa slight hit indeed. Still think it's worth it

master with boxfill ran 80 times:
```
MAXIMUM PCTAGE WAS: 120.243566445 TIME: 0.366565227509 AVERAGE: 0.393683287501
```

in this PR

```
MAXIMUM PCTAGE WAS: 114.226072659 TIME: 0.383887052536 AVERAGE: 0.404870161414
```

**Comment by danlipsa** [comment]
@doutriaux1 Go for it! +2

**Pull request merged**